### PR TITLE
Consumer

### LINKED ISSUE #94: Provide a reliable message publisher
A reliable publisher should:
* Queue up messages internally until it can successfully call basicPublish.
* Notify the caller using a robust ID (not delivery tag) when the message is confirmed by rabbit.
This is a layer above the existing abstraction in vertx-rabbitmq-client, but it's such common boiler plate (and non-trivial) that adding to vertx-rabbitmq-client would be useful.

Note that I am aiming to guarantee at least once delivery, there will be failure modes that result in messages being delivered more than once.
Also, clients that don't need at least once delivery guarantees should not use this publisher.

### PR DESCRIPTION
**Pull request by Yaytay**
Motivation:

Fixes #94 and all the changes requested.
Has a couple of new changes following stress testing.

Conformance:

Eclipse Contributor Agreement signed.

### CODE REVIEW

**Comment by vietj** [comment]
can you rebase this PR on top of the latest master ? there are conflicts

File: src/main/java/io/vertx/rabbitmq/RabbitMQConfirmation.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package io.vertx.rabbitmq;
+
+/**
+ *
+ * @author jtalbut
+ */
+public class RabbitMQConfirmation {
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQConfirmation.java]: this should be annotated with `@DataObject`

File: src/main/java/io/vertx/rabbitmq/RabbitMQConsumer.java [hunk 2]
```diff
@@ -66,8 +66,14 @@
    * <p>
    * The operation is asynchronous. When consumption will be stopped, you can by notified via {@link RabbitMQConsumer#endHandler(Handler)}
    */
-  void cancel();
+  Future<Void> cancel();
 
+  /**
+   * Return true if cancel() has been called.
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQConsumer.java]: {@code true}

File: src/main/java/io/vertx/rabbitmq/RabbitMQMessage.java [hunk 3]
```diff
@@ -45,4 +45,6 @@
    */
   @CacheReturn
   Integer messageCount();
+
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQMessage.java]: unecessary reformat

File: src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+   * Like {@link #basicPublish(String, String, BasicProperties, Buffer, Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  Future<Void> publish(String exchange, String routingKey, BasicProperties properties, Buffer body);
+  
+  /**
+   * Get the number of published, but not sent, messages.
+   * @return the number of published, but not sent, messages.
+   */
+  int getQueueSize();
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java]: `queueSize()` instead

File: src/main/java/io/vertx/rabbitmq/RabbitMQPublisherConfirmation.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package io.vertx.rabbitmq;
+
+import io.vertx.codegen.annotations.DataObject;
+
+/**
+ *
+ * @author njt
+ */
+@DataObject(generateConverter = true)
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQPublisherConfirmation.java]: why generating a converter if that's not used ?

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java [hunk 6]
```diff
@@ -4,46 +4,68 @@
...
+  @Override
+  public void handleConsumeOk(String consumerTag) {
+    super.handleConsumeOk(consumerTag);
+    log.info("Consumer tag is now " + consumerTag);
+  }
+
   @Override
   public void handleCancel(String consumerTag) {
-    log.debug("consumer has been cancelled unexpectedly");
+    log.info("consumer has been cancelled unexpectedly: " + consumerTag);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: don't change this

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java [hunk 7]
```diff
@@ -4,46 +4,68 @@
...
   @Override
   public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) {
     RabbitMQMessage msg = new RabbitMQMessageImpl(body, consumerTag, envelope, properties, null);
     this.handlerContext.runOnContext(v -> queue.handleMessage(msg));
   }
 
+  @Override
+  public void handleConsumeOk(String consumerTag) {
+    super.handleConsumeOk(consumerTag);
+    log.info("Consumer tag is now " + consumerTag);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: use debug

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java [hunk 8]
```diff
@@ -4,46 +4,68 @@
...
   @Override
   public void handleCancel(String consumerTag) {
-    log.debug("consumer has been cancelled unexpectedly");
+    log.info("consumer has been cancelled unexpectedly: " + consumerTag);
     queue.handleEnd();
   }
 
+  @Override
+  public void handleShutdownSignal(String consumerTag, ShutdownSignalException sig) {
+    log.info("consumer has been shutdown unexpectedly: " + consumerTag);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: use debug

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 9]
```diff
@@ -145,6 +174,51 @@ default void basicConsumer(String queue, Handler<AsyncResult<RabbitMQConsumer>>
...
 
+  /**
+   * Publish a message. Publishing to a non-existent exchange will result in a channel-level protocol exception,
+   * which closes the channel. Invocations of Channel#basicPublish will eventually block if a resource-driven alarm is in effect.
+   *
+   * @param deliveryTagHandler callback to capture the deliveryTag for this message.  Note that this will be called synchronously in the context of the client.
+   * @see com.rabbitmq.client.Channel#basicPublish(String, String, AMQP.BasicProperties, byte[])
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  void basicPublishWithDeliveryTag(String exchange, String routingKey, BasicProperties properties, Buffer body, @Nullable Handler<Long> deliveryTagHandler, Handler<AsyncResult<Void>> resultHandler);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: remove `deliveryTagHandler` and instead have resultHandler be a `Handler<List<Long>>`

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 10]
```diff
@@ -145,6 +174,51 @@ default void basicConsumer(String queue, Handler<AsyncResult<RabbitMQConsumer>>
...
+
+  /**
+   * Publish a message. Publishing to a non-existent exchange will result in a channel-level protocol exception,
+   * which closes the channel. Invocations of Channel#basicPublish will eventually block if a resource-driven alarm is in effect.
+   *
+   * @param deliveryTagHandler callback to capture the deliveryTag for this message.  Note that this will be called synchronously in the context of the client.
+   * @see com.rabbitmq.client.Channel#basicPublish(String, String, AMQP.BasicProperties, byte[])
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  Future<Void> basicPublishWithDeliveryTag(String exchange, String routingKey, BasicProperties properties, Buffer body, @Nullable Handler<Long> deliveryTagHandler);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: change `deliveryTagHandler` to be a resultHandler as a `Handler<AsyncResult<List<Long>>>`

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 9 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: The delivery tag needs to be known before the result is complete.
The confirmation may be sent before the resultHandler gets called.

I can expand the javadoc if that would make this clearer.

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java (hunk 7 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: @vietj Are you OK with me changing it to:
log.debug("consumer has been cancelled unexpectedly: " + consumerTag);
Knowing which consumer is shut down is very useful.

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 9 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: Added the comment:
```
   * The deliveryTagHandler will be called before the message is sent, which is necessary because the confirmation may arrive asynchronously before the resultHandler is called.
```

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 10 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: See above - the delivery tag may be needed before the result of the publish is known.

File: pom.xml [hunk 11]
```diff
@@ -13,8 +13,8 @@
   <version>4.0.0-SNAPSHOT</version>
 
   <properties>
-    <stack.version>4.0.0-SNAPSHOT</stack.version>
-    <rabbitmq.client.version>5.4.2</rabbitmq.client.version>
+    <stack.version>4.0.0-milestone4</stack.version>
```
Comment (vietj) [inline pom.xml:16]: why is this changed ?

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 9 above)
Comment (vietj) [reply src/main/java/io/vertx/rabbitmq/RabbitMQClient.java:189]: it can be known before, however that does not prevent the resultHandler to be called I think, i.e is there an interaction that can be triggered by the deliveryTagHandler before the resultHandler is called that will have any side effect on the result ?

File: src/main/java/io/vertx/rabbitmq/RabbitMQConsumer.java [hunk 12]
```diff
@@ -66,8 +66,14 @@
    * <p>
    * The operation is asynchronous. When consumption will be stopped, you can by notified via {@link RabbitMQConsumer#endHandler(Handler)}
    */
-  void cancel();
+  Future<Void> cancel();
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQConsumer.java:69]: if we do have this cancel method with a Future return then we need also to have an `cancel(Handler<AsyncResult<Void>>)` method alternative

File: src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+  
+  /**
+   * Like {@link #stop(Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<Void> stop();
+  
+  /**
+   * Undo the effects of calling {@link #stop(Handler)} so that publish may be called again. 
+   */
+  void restart();
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java:74]: what does restart do if stop() is not called ?

File: src/main/java/io/vertx/rabbitmq/impl/RabbitMQPublisherImpl.java [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+  
+  private static final Logger log = LoggerFactory.getLogger(RabbitMQPublisherImpl.class);
+  
+  private final Vertx vertx;
+  private final RabbitMQClient client;
+  private final InboundBuffer<RabbitMQPublisherConfirmation> confirmations;
+  private final Context context;
+  private final RabbitMQPublisherOptions options;
+  
+  private final Deque<MessageDetails> awaitingAck = new ArrayDeque<>();
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/impl/RabbitMQPublisherImpl.java:53]: rather `pendingAcks`

File: pom.xml (hunk 11 above)
Comment (Yaytay) [reply pom.xml:16]: Stupid error on my part because I'm using the compiled jar in my projects - shouldn't have made it to the commit.

File: src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java (hunk 13 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/RabbitMQPublisher.java:74]: It's harmless to call restart when stop hasn't been called (it'll pause and resume the stream, that's all).
However, if restart is called when stop has been called but has not yet completed then the stop will never complete.
I've added a comment to say that, do you think it needs more?

File: src/main/java/io/vertx/rabbitmq/impl/RabbitMQPublisherImpl.java (hunk 14 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/impl/RabbitMQPublisherImpl.java:53]: Changed

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java (hunk 6 above)
Comment (Yaytay) [reply src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: I've changed back to debug, but left the consumerTag on there.

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java (hunk 7 above)
Comment (vietj) [reply src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: yes it is fine

File: src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java (hunk 6 above)
Comment (vietj) [reply src/main/java/io/vertx/rabbitmq/impl/QueueConsumerHandler.java]: sure, it was about info

**Comment by vietj** [comment]
can you rebase due to conflicts ? also we need documentation in `index.adoc`

**Comment by vietj** [comment]
can you add missing documentation ?

**Comment by Yaytay** [comment]
Hi Julien,

I've added to index.adoc and the examples file.
I hope it comes out OK, it's my first use of Ascii Docs.

I also had to tweak one of the unit tests because it was deadlocking, 
but that was just a case of bringing it into line with master.

Jim

**Comment by Yaytay** [comment]
Hi Julien,

Have you had a chance to look at the documentation I added?

Thanks

Jim
